File: mage/src/main/java/mil/nga/giat/mage/location/LocationReportingService.java
package mil.nga.giat.mage.location;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.arch.lifecycle.LifecycleService;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Observer;
import android.content.Intent;
import android.content.SharedPreferences;
import android.location.Location;
import android.location.LocationManager;
import android.os.Build;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;
import android.util.Log;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import mil.nga.giat.mage.MAGE;
import mil.nga.giat.mage.R;

public class LocationReportingService extends LifecycleService implements
        Observer<Location>,
        LocationSaveTask.LocationDatabaseListener,
        LocationPushTask.LocationSyncListener,
        SharedPreferences.OnSharedPreferenceChangeListener {

    private static final String LOG_NAME = LocationReportingService.class.getName();
    public static final String NOTIFICATION_CHANNEL_ID = "mil.nga.mage.LOCATION_NOTIFICATION_CHANNEL";

    private static final int NOTIFICATION_ID = 500;

    public static final Executor SAVE_EXECUTOR = Executors.newSingleThreadExecutor();
    public static final Executor PUSH_EXECUTOR = new ThreadPoolExecutor(1, 1, 1L, TimeUnit.SECONDS, new SynchronousQueue<Runnable>(), new ThreadPoolExecutor.DiscardPolicy());

    private SharedPreferences preferences;
    private LiveData<Location> locationLiveData;

    private boolean shouldReportLocation;
    private long locationPushFrequency;
    private long oldestLocationTime = 0;

    @Override
    public void onCreate() {
        super.onCreate();

        Log.v(LOG_NAME, "onCreate LocationService");

        preferences = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
        preferences.registerOnSharedPreferenceChangeListener(this);

        locationPushFrequency = getLocationPushFrequency();
        shouldReportLocation = getShouldReportLocation();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationManager notificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
            NotificationChannel channel = new NotificationChannel(NOTIFICATION_CHANNEL_ID,"MAGE", NotificationManager.IMPORTANCE_MIN);
            channel.setShowBadge(true);
            notificationManager.createNotificationChannel(channel);
        }

        Notification notification = new NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID)
                .setContentTitle("MAGE Location Service")
                .setContentText("MAGE is currently reporting your location.")
                .setSmallIcon(R.drawable.ic_place_white_24dp)
                .setGroup(MAGE.MAGE_NOTIFICATION_GROUP)
                .build();

        startForeground(NOTIFICATION_ID, notification);

        locationLiveData = ((MAGE) getApplication()).getLocationLiveData();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        super.onStartCommand(intent, flags, startId);

        Log.v(LOG_NAME, "onStartCommand received from service, this service has been restarted");
        locationLiveData.observe(this, this);

        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        Log.v(LOG_NAME, "onDestroy received from service, this service is being destroyed");

        locationLiveData.removeObserver(this);

        preferences.unregisterOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onChanged(@Nullable Location location) {
        if (shouldReportLocation && location.getProvider().equals(LocationManager.GPS_PROVIDER)) {
            Log.v(LOG_NAME, "GPS location changed");
            new LocationSaveTask(getApplicationContext(), LocationReportingService.this).executeOnExecutor(SAVE_EXECUTOR, location);
        }
    }

    @Override
    public void onSaveComplete(Location location) {
        if (location.getTime() - oldestLocationTime > locationPushFrequency) {
            new LocationPushTask(getApplicationContext(), this).executeOnExecutor(PUSH_EXECUTOR);
        }

        if (oldestLocationTime == 0) {
            oldestLocationTime = location.getTime();
        }
    }

    @Override
    public void onSyncComplete(Boolean status) {
        if (status) {
            oldestLocationTime = 0;
        }
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (key.equalsIgnoreCase(getString(R.string.reportLocationKey))) {
            shouldReportLocation = getShouldReportLocation();
            Log.d(LOG_NAME, "Report location changed " + shouldReportLocation);
        } else if (key.equalsIgnoreCase(getString(R.string.locationPushFrequencyKey))) {
            locationPushFrequency = getLocationPushFrequency();
            Log.d(LOG_NAME, "Location push frequency changed " + locationPushFrequency);
        }
    }

    private long getLocationPushFrequency() {
        return preferences.getInt(getString(R.string.locationPushFrequencyKey), getResources().getInteger(R.integer.locationPushFrequencyDefaultValue));
    }

    private boolean getShouldReportLocation() {
        return preferences.getBoolean(getString(R.string.reportLocationKey), getResources().getBoolean(R.bool.reportLocationDefaultValue));
    }
}

File: mage/src/main/java/mil/nga/giat/mage/location/LocationReportingService.kt
package mil.nga.giat.mage.location

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.location.Location
import android.location.LocationManager
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.lifecycle.LifecycleService
import androidx.lifecycle.lifecycleScope
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.launch
import mil.nga.giat.mage.MageApplication
import mil.nga.giat.mage.R
import mil.nga.giat.mage.data.repository.location.LocationRepository
import mil.nga.giat.mage.login.LoginActivity
import java.util.concurrent.atomic.AtomicBoolean
import javax.inject.Inject

@AndroidEntryPoint
open class LocationReportingService : LifecycleService(), SharedPreferences.OnSharedPreferenceChangeListener {

    @Inject lateinit var locationProvider: LocationProvider
    @Inject lateinit var locationRepository: LocationRepository
    @Inject lateinit var locationAccess: LocationAccess
    @Inject lateinit var preferences: SharedPreferences
    @Inject lateinit var mageApp: MageApplication

    private var shouldReportLocation: Boolean = false
    private var locationPushFrequency: Long = 0
    private var locationTimeForLastPush: Long = 0
    private var locationUpdatesJob: Job? = null
    private var isFirstLocationInSession = true
    private val isPushing = AtomicBoolean(false)

    companion object {
        private val LOG_NAME = LocationReportingService::class.java.name

        const val NOTIFICATION_ID = 500
        const val NOTIFICATION_CHANNEL_ID = "mil.nga.mage.LOCATION_NOTIFICATION_CHANNEL"
    }

    override fun onCreate() {
        super.onCreate()

        // If the user disables the location permission from settings, MAGE will be restarted, including this service (Service.START_STICKY)
        // Check for location permission here as it may have been disabled, if so stop the service.
        if (locationAccess.isLocationDenied()) {
            stopForeground(true)
            return
        }

        preferences.registerOnSharedPreferenceChangeListener(this)

        locationPushFrequency = getLocationPushFrequency()
        shouldReportLocation = mageApp.shouldReportLocation()

        val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val notificationChannel = NotificationChannel(NOTIFICATION_CHANNEL_ID, "MAGE", NotificationManager.IMPORTANCE_MIN)
        notificationChannel.setShowBadge(true)
        notificationManager.createNotificationChannel(notificationChannel)

        val intent = Intent(applicationContext, LoginActivity::class.java)
        intent.putExtra("LOGOUT", true)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        val pendingIntent = PendingIntent.getActivity(applicationContext, 1, intent, PendingIntent.FLAG_IMMUTABLE)

        val notification = NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID)
            .setContentTitle("MAGE Location Service")
            .setContentText("MAGE is currently reporting your location.")
            .setSmallIcon(R.drawable.ic_place_white_24dp)
            .setGroup(MageApplication.MAGE_NOTIFICATION_GROUP)
            .addAction(R.drawable.ic_power_settings_new_white_24dp, "Logout", pendingIntent)
            .build()

        startForeground(NOTIFICATION_ID, notification)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        super.onStartCommand(intent, flags, startId)

        if (locationUpdatesJob == null) {
            locationUpdatesJob = lifecycleScope.launch {
                locationProvider.locationUpdates.filterNotNull().collect { location ->
                    if (shouldReportLocation && location.provider == LocationManager.GPS_PROVIDER) {
                        Log.v(LOG_NAME, "GPS location changed")

                        launch {
                            locationRepository.saveLocation(location)
                            pushLocations(location)
                        }
                    }
                }
            }
        }

        return START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        try {
            locationUpdatesJob?.cancel()
            preferences.unregisterOnSharedPreferenceChangeListener(this)
        } catch (e: Exception) {
            Log.d(LOG_NAME, "Error shutting down service: " + e.message)
        }
    }

    private suspend fun pushLocations(location: Location) {
        if (isFirstLocationInSession ||
            (location.time - locationTimeForLastPush > locationPushFrequency) ||
            !locationAccess.isPreciseLocationGranted()) {
            if (isPushing.compareAndSet(false, true)) {
                try {
                    val success = locationRepository.pushLocations()
                    if (success) {
                        isFirstLocationInSession = false
                        locationTimeForLastPush = location.time
                    }
                } finally {
                    isPushing.set(false)
                }
            }
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        if (key.equals(getString(R.string.reportLocationKey), ignoreCase = true)) {
            shouldReportLocation = mageApp.shouldReportLocation()
            Log.d(LOG_NAME, "Report location changed $shouldReportLocation")
        } else if (key.equals(getString(R.string.locationPushFrequencyKey), ignoreCase = true)) {
            locationPushFrequency = getLocationPushFrequency()
            Log.d(LOG_NAME, "Location push frequency changed $locationPushFrequency")
        }
    }

    private fun getLocationPushFrequency(): Long {
        return preferences.getInt(getString(R.string.locationPushFrequencyKey), resources.getInteger(R.integer.locationPushFrequencyDefaultValue)).toLong()
    }
}
